[GStreamer] fix deadlock when seeking

gdesmott opened the pull request:
#### 329f047fc4263d38c8102e9378148b32a52ff529
<pre>
[GStreamer] fix deadlock when seeking
<a href="https://bugs.webkit.org/show_bug.cgi?id=292527">https://bugs.webkit.org/show_bug.cgi?id=292527</a>

Reviewed by Philippe Normand.

Prevent deadlock in the main thread between the STREAM lock and downloadbuffer waiting for
upstream data when running with WEBKIT_GST_USE_PLAYBIN3=1

Fix this by sending the seek event from another thread when sent because
of a state change.

Also call gst_bin_recalculate_latency() from another thread for the same
reason.

Canonical link: <a href="https://commits.webkit.org/294507@main">https://commits.webkit.org/294507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25f8d9dadc33df27b8e931e31d83177417fc1fc5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102112 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21779 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12095 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107271 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52747 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77720 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34713 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92190 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58055 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10215 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52105 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86761 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10288 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109645 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21561 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86690 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88394 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86273 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21942 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31075 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8795 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/23474 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29172 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34467 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28983 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32306 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30542 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -955,7 +960,7 @@ void connectSimpleBusMessageCallback(GstElement* pipeline, Function<void(GstMess
             // for one reason or another a new live element is added or
             // removed from the pipeline.
-            gst_bin_recalculate_latency(GST_BIN_CAST(pipeline.get()));
+            gst_element_call_async (pipeline.get(), recalculate_latency_async, NULL, NULL);
```
Was this also triggering deadlocks? 

BTW s/NULL/nullptr in the whole patch please :)

gdesmott replied:
> Was this also triggering deadlocks?

Yes, same as the first one: main thread being blocked.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -912,6 +912,11 @@ struct MessageBusData {
 };
 WEBKIT_DEFINE_ASYNC_DATA_STRUCT(MessageBusData)
 
+static void recalculate_latency_async (GstElement * pipeline, gpointer user_data) {
```
We use camelCase for functions

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -600,8 +605,16 @@ bool MediaPlayerPrivateGStreamer::doSeek(const SeekTarget& target, float rate)
-    return gst_element_seek(m_pipeline.get(), rate, GST_FORMAT_TIME, m_seekFlags, GST_SEEK_TYPE_SET, seekStart, GST_SEEK_TYPE_SET, seekStop);
+
+    if (isAsync) {
+        gst_element_call_async(m_pipeline.get(), seek_async, event, NULL);
```
Maybe instead of a static func use a closure:

```
gst_element_call_async(m_pipeline.get(), reinterpret_cast<GstElementCallAsyncFunc>(+[](GstElement*, gpointer) {
    ...
}), event, nullptr);
```

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -955,7 +961,7 @@ void connectSimpleBusMessageCallback(GstElement* pipeline, Function<void(GstMess
             // for one reason or another a new live element is added or
             // removed from the pipeline.
-            gst_bin_recalculate_latency(GST_BIN_CAST(pipeline.get()));
+            gst_element_call_async(pipeline.get(), recalculateLatencyAsync, NULL, NULL);
```
nullptr

gdesmott committed:
[GStreamer] fix deadlock when seeking
https://bugs.webkit.org/show_bug.cgi?id=292527

Reviewed by Philippe Normand.

Prevent deadlock in the main thread between the STREAM lock and downloadbuffer waiting for
upstream data when running with WEBKIT_GST_USE_PLAYBIN3=1

Fix this by sending the seek event from another thread when sent because
of a state change.

Also call gst_bin_recalculate_latency() from another thread for the same
reason.

Canonical link: https://commits.webkit.org/294507@main
